leetcode/src/61.c

```diff
@@ -0,0 +1,48 @@
+/**
+* Given the head of a linked list, rotate the list to the right by k places.
+*
+* This code takes a singly-linked list head and an integer k as input and rotates the list to the right by k positions.
+* The function returns a pointer to the new head of the rotated list.
+
+* The first while loop iterates over the linked list to count the number of nodes. This loop takes O(n) time
+* The next if statement takes O(1) time.
+* The second while loop rotates the linked list by k positions. The loop runs k times and performs a constant amount of work in each iteration, including traversing the linked list to find the last node, setting the new head, and updating the pointers. The loop takes O(k) time.
+* Overall time complexity of the code is O(n + k).
+*
+* The code uses only a constant amount of extra memory, regardless of the size of the input. Therefore, the space complexity of the code is O(1).
+
+* Definition for singly-linked list.
+* struct ListNode {
+* int val;
+* struct ListNode *next;
+* };
+*/
+
+struct ListNode* rotateRight(struct ListNode* head, int k) {
+
+    struct ListNode *ptr = head, *prev =NULL;
+    int count = 0;
+
+    while(ptr) {
+        ptr = ptr->next;
+        count++;
+    }
+    if(count!=0)
+        k = k%count;
+    while(k--) {
+        ptr = head;
+        if(ptr==NULL)
+            return NULL;
+        if(ptr->next==NULL)
+            return ptr;
+        while(ptr->next) {
```

Comment:
Are you sure that the Runtime complexity is O(n + k)? I see the outer loop is O(k) and the inner loop goes throught all the list so, If I'm right it's the Runtime O(n * k), no?

Reply:
> Are you sure that the Runtime complexity is O(n + k)? I see the outer loop is O(k) and the inner loop goes throught all the list so, If I'm right it's the Runtime O(n * k), no?

I think you are right that it's not O(n + k), but I dissent that the outer loop is really O(k mod n), so O(n * k) would be the worst case time complexity (When k = n - 1).